File: archive/cdk-java/Pr2376/lib/src/main/java/app/constructs/VpcConstruct.java
package app.constructs;

import software.amazon.awscdk.services.ec2.CfnEIP;
import software.amazon.awscdk.services.ec2.CfnInternetGateway;
import software.amazon.awscdk.services.ec2.CfnNatGateway;
import software.amazon.awscdk.services.ec2.CfnRoute;
import software.amazon.awscdk.services.ec2.CfnRouteTable;
import software.amazon.awscdk.services.ec2.CfnSubnetRouteTableAssociation;
import software.amazon.awscdk.services.ec2.CfnVPCGatewayAttachment;
import software.amazon.awscdk.services.ec2.IpAddresses;
import software.amazon.awscdk.services.ec2.Subnet;
import software.amazon.awscdk.services.ec2.Vpc;
import software.constructs.Construct;

public class VpcConstruct extends Construct {
    private final Vpc vpc;
    private final Subnet publicSubnet;

    private final Subnet privateSubnet;

    public VpcConstruct(final Construct scope, final String id, final String vpcCidr) {
        super(scope, id);

        // Create VPC
        this.vpc = Vpc.Builder.create(this, "Vpc")
                .ipAddresses(IpAddresses.cidr(vpcCidr))
                .maxAzs(1) // Use single AZ for simplicity
                .enableDnsHostnames(true)
                .enableDnsSupport(true)
                .subnetConfiguration(java.util.List.of())
                .natGateways(0) // We'll create NAT Gateway manually
                .build();

        // Create Internet Gateway (automatically created with VPC, but we reference it)
        CfnInternetGateway igw = CfnInternetGateway.Builder.create(this, "InternetGateway")
                .build();

        CfnVPCGatewayAttachment.Builder.create(this, "VpcGatewayAttachment")
                .vpcId(vpc.getVpcId())
                .internetGatewayId(igw.getRef())
                .build();

        // Get the first availability zone
        String availabilityZone = vpc.getAvailabilityZones().get(0);

        // Create public subnet
        this.publicSubnet = Subnet.Builder.create(this, "PublicSubnet")
                .vpcId(vpc.getVpcId())
                .cidrBlock(calculateSubnetCidr(vpcCidr, 0)) // First subnet
                .availabilityZone(availabilityZone)
                .mapPublicIpOnLaunch(true)
                .build();

        // Create private subnet
        // Second subnet
        this.privateSubnet = Subnet.Builder.create(this, "PrivateSubnet")
                .vpcId(vpc.getVpcId())
                .cidrBlock(calculateSubnetCidr(vpcCidr, 1)) // Second subnet
                .availabilityZone(availabilityZone)
                .mapPublicIpOnLaunch(false)
                .build();

        // Create Elastic IP for NAT Gateway
        CfnEIP natEip = CfnEIP.Builder.create(this, "NatEip")
                .domain("vpc")
                .build();

        // Create NAT Gateway in public subnet
        CfnNatGateway natGateway = CfnNatGateway.Builder.create(this, "NatGateway")
                .subnetId(publicSubnet.getSubnetId())
                .allocationId(natEip.getAttrAllocationId())
                .build();

        // Create route table for public subnet
        CfnRouteTable publicRouteTable = CfnRouteTable.Builder.create(this, "PublicRouteTable")
                .vpcId(vpc.getVpcId())
                .build();

        // Add route to Internet Gateway for public subnet
        CfnRoute.Builder.create(this, "PublicRoute")
                .routeTableId(publicRouteTable.getAttrRouteTableId())
                .destinationCidrBlock("0.0.0.0/0")
                .gatewayId(igw.getAttrInternetGatewayId())
                .build();

        // Associate public subnet with public route table
        CfnSubnetRouteTableAssociation.Builder.create(this, "PublicSubnetRouteTableAssociation")
                .subnetId(publicSubnet.getSubnetId())
                .routeTableId(publicRouteTable.getAttrRouteTableId())
                .build();

        // Create route table for private subnet
        CfnRouteTable privateRouteTable = CfnRouteTable.Builder.create(this, "PrivateRouteTable")
                .vpcId(vpc.getVpcId())
                .build();

        // Add route to NAT Gateway for private subnet
        CfnRoute.Builder.create(this, "PrivateRoute")
                .routeTableId(privateRouteTable.getAttrRouteTableId())
                .destinationCidrBlock("0.0.0.0/0")
                .natGatewayId(natGateway.getAttrNatGatewayId())
                .build();

        // Associate private subnet with private route table
        CfnSubnetRouteTableAssociation.Builder.create(this, "PrivateSubnetRouteTableAssociation")
                .subnetId(privateSubnet.getSubnetId())
                .routeTableId(privateRouteTable.getAttrRouteTableId())
                .build();
    }

    private String calculateSubnetCidr(final String vpcCidr, final int subnetIndex) {
        String[] parts = vpcCidr.split("\\.");
        return parts[0] + "." + parts[1] + "." + subnetIndex + ".0/24";
    }

    public Vpc getVpc() {
        return vpc;
    }

    public Subnet getPublicSubnet() {
        return publicSubnet;
    }

    public Subnet getPrivateSubnet() {
        return privateSubnet;
    }
}

File: archive/cdktf-ts/Pr7795/lib/constructs/vpc-construct.ts
import { Construct } from 'constructs';
import { Vpc } from '@cdktf/provider-aws/lib/vpc';
import { Subnet } from '@cdktf/provider-aws/lib/subnet';
import { InternetGateway } from '@cdktf/provider-aws/lib/internet-gateway';
import { RouteTable } from '@cdktf/provider-aws/lib/route-table';
import { RouteTableAssociation } from '@cdktf/provider-aws/lib/route-table-association';
import { Route } from '@cdktf/provider-aws/lib/route';

export interface VpcConstructProps {
  environmentSuffix: string;
  cidrBlock: string;
  availabilityZones: string[];
  tags?: Record<string, string>;
}

export class VpcConstruct extends Construct {
  public readonly vpc: Vpc;
  public readonly publicSubnets: Subnet[];
  public readonly privateSubnets: Subnet[];
  public readonly internetGateway: InternetGateway;

  constructor(scope: Construct, id: string, props: VpcConstructProps) {
    super(scope, id);

    const {
      environmentSuffix,
      cidrBlock,
      availabilityZones,
      tags = {},
    } = props;

    // Create VPC
    this.vpc = new Vpc(this, 'vpc', {
      cidrBlock,
      enableDnsHostnames: true,
      enableDnsSupport: true,
      tags: {
        Name: `vpc-${environmentSuffix}`,
        ...tags,
      },
    });

    // Create Internet Gateway
    this.internetGateway = new InternetGateway(this, 'igw', {
      vpcId: this.vpc.id,
      tags: {
        Name: `igw-${environmentSuffix}`,
        ...tags,
      },
    });

    // Create public and private subnets in each AZ
    this.publicSubnets = [];
    this.privateSubnets = [];

    availabilityZones.forEach((az, index) => {
      // Public subnet
      const publicSubnet = new Subnet(this, `public-subnet-${index}`, {
        vpcId: this.vpc.id,
        cidrBlock: this.calculateSubnetCidr(cidrBlock, index * 2),
        availabilityZone: az,
        mapPublicIpOnLaunch: true,
        tags: {
          Name: `public-subnet-${index}-${environmentSuffix}`,
          Type: 'public',
          ...tags,
        },
      });
      this.publicSubnets.push(publicSubnet);

      // Private subnet
      const privateSubnet = new Subnet(this, `private-subnet-${index}`, {
        vpcId: this.vpc.id,
        cidrBlock: this.calculateSubnetCidr(cidrBlock, index * 2 + 1),
        availabilityZone: az,
        mapPublicIpOnLaunch: false,
        tags: {
          Name: `private-subnet-${index}-${environmentSuffix}`,
          Type: 'private',
          ...tags,
        },
      });
      this.privateSubnets.push(privateSubnet);
    });

    // Create route table for public subnets
    const publicRouteTable = new RouteTable(this, 'public-rt', {
      vpcId: this.vpc.id,
      tags: {
        Name: `public-rt-${environmentSuffix}`,
        ...tags,
      },
    });

    // Add route to Internet Gateway
    new Route(this, 'public-route', {
      routeTableId: publicRouteTable.id,
      destinationCidrBlock: '0.0.0.0/0',
      gatewayId: this.internetGateway.id,
    });

    // Associate public subnets with public route table
    this.publicSubnets.forEach((subnet, index) => {
      new RouteTableAssociation(this, `public-rt-assoc-${index}`, {
        subnetId: subnet.id,
        routeTableId: publicRouteTable.id,
      });
    });

    // Create route table for private subnets
    const privateRouteTable = new RouteTable(this, 'private-rt', {
      vpcId: this.vpc.id,
      tags: {
        Name: `private-rt-${environmentSuffix}`,
        ...tags,
      },
    });

    // Associate private subnets with private route table
    this.privateSubnets.forEach((subnet, index) => {
      new RouteTableAssociation(this, `private-rt-assoc-${index}`, {
        subnetId: subnet.id,
        routeTableId: privateRouteTable.id,
      });
    });
  }

  private calculateSubnetCidr(vpcCidr: string, subnetIndex: number): string {
    const [network] = vpcCidr.split('/');
    const [a, b] = network.split('.').map(Number);
    const newC = subnetIndex * 16;
    return `${a}.${b}.${newC}.0/20`;
  }
}
